qutebrowser/completion/models/sql.py

```diff
@@ -0,0 +1,302 @@
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The base completion model for completion in the command line.
+
+Module attributes:
+    Role: An enum of user defined model roles.
+"""
+
+import re
+
+from PyQt5.QtCore import Qt, QModelIndex, QAbstractItemModel
+from PyQt5.QtSql import QSqlTableModel, QSqlDatabase, QSqlQuery
+
+from qutebrowser.utils import usertypes, log
+
+
+Role = usertypes.enum('Role', ['sort'], start=Qt.UserRole, is_int=True)
+
+
+def init(path):
+    """Initialize the SQL completion module.
+
+    Args:
+        path: Path to the completion database.
+    """
+    database = QSqlDatabase.addDatabase('QSQLITE')
+    database.setDatabaseName(path)
+    if not database.open():
+        raise SqlException("Failed to open {}".format(path))
+
+
+def close():
+    """Close the SQL connection."""
+    QSqlDatabase.removeDatabase(QSqlDatabase.database().connectionName())
+
+
+def _run_query(querystr):
+    """Run the given SQL query string on the database."""
+    log.completion.debug('Running SQL query: "{}"'.format(querystr))
+    database = QSqlDatabase.database('completions')
+    query = QSqlQuery(database)
+    if not query.exec_(querystr):
+        raise SqlException('Failed to exec query "{}": "{}"'
+                           .format(querystr, query.lastError().text()))
+
+
+class CompletionCategory(QSqlTableModel):
+
+    """Wraps a sql table providing data for a completion category."""
+
+    def new_item(self, name, desc='', misc=None, sort=None):
+        """Add a new item to a category.
+
+        Fails if an item with the same name exists.
+
+        Args:
+            name: Data for the first column.
+            desc: Data for the second column.
+            misc: Data for the third column.
+            sort: Optional data for the sorting column (not visible).
+        """
+        record = self.record()
+        record.setValue('name', name)
+        record.setValue('desc', desc)
+        record.setValue('misc', misc)
+        record.setValue('sort', sort)
+        if not self.insertRecord(-1, record):
+            raise SqlException("Failed to insert '{}': {}"
+                               .format(name, self.lastError().text()))
+        self.select()
+
+    def remove_item(self, key):
+        """Remove an item from a category.
+
+        Fails if the item is not found.
+
+        Args:
+            key: The primary key value of the item to remove.
+        """
+        field = self.primaryKey().fieldName(0)
+        _run_query("DELETE FROM {} where {} = '{}'"
+                   .format(self.tableName(), field, key))
```

Comment:
We really shouldn't use string formatting for SQL queries - QSQLQuery supports [binding values](http://doc.qt.io/qt-5/qsqlquery.html#approaches-to-binding-values), and `_run_query` can probably abstract that nicely using keyword arguments.